docs/reference/remap/append.cue

```diff
@@ -0,0 +1,39 @@
+package metadata
+
+remap: functions: append: {
+	arguments: [
+		{
+			name:        "value"
+			description: "The array"
+			required:    true
+			type: ["array"]
+		},
+		{
+			name:        "item"
+			description: "The item to append"
+			required:    true
+			type: ["any"]
+		},
+	]
+	return: ["array"]
```

Comment:
This might warrant a larger discussion, but I think it would be clearer that `append` modifies the array in-place if it didn't return anything. The fact that it returns the modified array could lead users to believe it is returning a new array that is a copy of the old array plus the new element.

Reply:
I do know that Jean et al would eventually like to make `del`, for example, return either the deleted value or `null` if the field doesn't exist. Returning the new array would enable you to do things like `unique(append(.fruits, "mango"))` (mentioned in my other comment) and feels more consistent with *some* programming languages (Go et al) but not others (Rust et al). I don't have strong feelings there, especially on this first iteration. Happy to defer.

Reply:
I think Go is a bit special because it actually returns a new "struct" representing the array. _Usually_ this is pointing to the same underlying memory, but it'll actually be different if it had to extend the array. Consider that Go's in-place sorting routines do not return the sorted array. I think this makes it clearer that it is happening in-place.

That being said, maybe I'm in the minority here. I know Ruby returns the same array to make it easier to chain though they also have `sort` create a copy while `sort!` sorts in-place.

I guess the most important thing is consistency in VRL. We can choose to always return the modified value, even if it happens in-place, and hopefully that consistency will make it easy to reason about.

Reply:
I would update the docs to make it clear that the array itself is modified though, the current wording:

> Appends the specified item to an array and returns the new array

Might lead people to imagine it returns a new array. Specifically the word `new` there. Maybe:

> Appends the specified item to an array and returns the modified array

Reply:
Ah, I see what you're saying now. Would "resulting array" be a bit more on the money?

Reply:
I tend to agree with @jszwedko, returning an array is very confusing:

```
.original = [1,2,3]
.modified = append(.original, 4)
.original
=> [1,2,3,4]
```

That is an easy gotcha. Given that the function is operating on a _value_, it should return a copy and not modify the original value. `del` is a special case in that it is operating the `event`, not a value, so there is nothing to return.

Reply:
I've verified that this works as expected:

```
$ .original = [1, 2, 3]

$ append(.original, 4)
[1, 2, 3, 4]

$ .original
[1, 2, 3]

$ .modified = append(.original, 4)
[1, 2, 3, 4]

$ .modified
[1, 2, 3, 4]

$ .original
[1, 2, 3]
```

Reply:
It does appear you are correct! I can't find where the `clone` is actually happening for the `Vec<>` though. Do you know?

```
        let mut list = self.value.execute(state, object)?.try_array()?;
        let item = self.item.execute(state, object)?;

        list.push(item);

        Ok(list.into())
```

Makes it seem like it'd be modifying in-place. I'm assuming it is cloned at some level above that, but I haven't been able to track it down yet.

Could we modify the tests to assert that the original is unchanged?

On a side-note, it looks like the added behavior test is causing failures:

```
Dec 30 14:39:29.589 ERROR vector::unit_test: Failed to execute tests/behavior/transforms/remap.toml tests:                                                                                     
data did not match any variant of untagged enum TestInputValue for key `tests.input.log_fields.fruits` at line 1476 column 5. 
```

I'm guessing it's because the condition is comparing with an array which it doesn't seem like we are doing anywhere else, but the enum does seem like it should allow for this:

https://github.com/timberio/vector/blob/58d6e8ffa9f1fe7f5fdf31359435c8bc9f9e26c2/src/conditions/check_fields.rs#L20

Reply:
@jszwedko This invocation of `execute` on the `value` only *extracts* the array. Compare with the `del` function logic:

```rust
self.paths
    .iter()
    .try_for_each(|path| object.remove(path.as_ref(), false))?; // Object is manipulated
```

Reply:
True, I'm probably missing something in the plumbing of `remap-lang`. I would have assumed that complex types like the `Vec` and `BTreeMap` would have been accessed by reference to avoid unnecessary clones (though we do want the clone here).

Reply:
> I'm probably missing something in the plumbing of `remap-lang`.

You're not missing anything 😄. We opted to accept the clone, as a trade-off to keep the initial implementation simpler.

Since we're using the `bytes` crate, the cloning is actually relatively cheap, since it uses reference-counting internally, and other most other value types implement copy, and are thus cheap to clone (cloning a map is probably most expensive, since we're storing the keys as strings on the heap).

Reply:
Aha :thinking: thanks for the explanation!

@lucperkins I think it'd still be nice to add a test that asserts the original array was unchanged in case we change the behavior of always cloning values and need to add an explicit clone here.

Reply:
> Every VRL function except `del` is a "pure" function that generates a value but doesn't modify the object.

`merge` is another one. `merge(., .foo)` will merge `.foo` into the root.